drgn_tools/locking.py

```diff
@@ -85,3 +229,238 @@ def for_each_mutex_waiter(prog: Program, mutex: Object) -> Iterable[Object]:
         prog.type("struct mutex_waiter"), mutex.wait_list.address_of_(), "list"
     ):
         yield waiter.task
+
+
+######################################
+# rwsem
+######################################
+
+# Masks for rwsem.count
+_RWSEM_WRITER_LOCKED = 1 << 0
+_RWSEM_FLAG_WAITERS = 1 << 1
+_RWSEM_FLAG_HANDOFF = 1 << 2
+# Bits 8-62(i.e. 55 bits of counter indicate number of current readers that hold the lock)
+_RWSEM_READER_MASK = 0x7FFFFFFFFFFFFF00  # Bits 8-62 - 55-bit reader count
+_RWSEM_WRITER_MASK = 1 << 0
+_RWSEM_READER_SHIFT = 8
+
+# Masks for rwsem.owner
+_RWSEM_READER_OWNED = 1 << 0
+_RWSEM_ANONYMOUSLY_OWNED = 1 << 0  # For old kernels
+_RWSEM_RD_NONSPINNABLE = 1 << 1
+_RWSEM_WR_NONSPINNABLE = 1 << 2
+
+# Linux kernel commit 617f3ef95177840c77f59c2aec1029d27d5547d6 ("locking/rwsem:
+# Remove reader optimistic spinning") (in v5.11) removed optimistic spinning for
+# readers and hence left one bit to check for spinnable
+_RWSEM_NONSPINNABLE = 1 << 1
+
+
+def rwsem_has_spinner(rwsem: Object) -> bool:
+    """
+    Check if rwsem has optimistic spinners or not
+
+    :param rwsem: ``struct rw_semaphore *``
+    :returns: True if rwsem has optimistic spinners, False otherwise.
+    """
+    return osq_is_locked(rwsem.osq.address_of_())
+
+
+def for_each_rwsem_waiter_entity(rwsem: Object) -> Iterable[Object]:
+    """
+    Find rwsem_waiter(s) for given rwsem
+
+    :param rwsem: ``struct rw_semaphore *``
+    :returns: Iterator of ``struct rwsem_waiter``
+    """
+
+    for waiter in list_for_each_entry(
+        "struct rwsem_waiter", rwsem.wait_list.address_of_(), "list"
+    ):
+        yield waiter
+
+
+def get_rwsem_owner(rwsem: Object) -> Object:
+    """
+    Find owner of  given rwsem
+
+    :param rwsem: ``struct rw_semaphore *``
+    :returns: ``struct task_struct *`` if owner can be found, NULL otherwise
+    """
+    prog = rwsem.prog_
+    if not rwsem.count.counter.value_():
+        print("rwsem is free.")
+        return NULL(prog, "struct task_struct *")
+
+    if is_rwsem_writer_owned(rwsem):
+        if (rwsem.owner.type_.type_name() != "atomic_long_t") and (
+            rwsem.owner.value_() & _RWSEM_ANONYMOUSLY_OWNED
+        ):
+            print("rwsem is owned by anonymous writer")
```

Comment:
The idea behind keeping these prints was that when needed we can check owner of any standalone or embedded rwsem , from drgn shell or cli, using this helper and if the owner can't be obtained because of either of the above reasons, we get to know the reason. I have found this helpful when working side by side with crash and drgn , where sometime I locate a rwsem in crash and want to know who owns it. This is why I intend to keep these. Please let me know your thoughts.

Reply:
I see. I agree that this makes most sense for doing interactive debugging, if you directly call the function, then getting the message printed on the command line makes sense.

If this is used as part of a more complex helper, then it may be a bit confusing to see a message printed out which is not part of the helper. For example, suppose we had a helper that printed a table of rwsems. Then the printout wouldn't be very helpful, because this line gets printed in between each row of the table.

What if we could make a slightly more complex return type, that expresses everything. For example:

```python
import collections
import enum

class RwsemStateCode(enum.Enum):
    UNLOCKED = "unlocked"
    READER_OWNED = "reader owned"
    WRITER_OWNED = "writer owned"
    WRITER_ANONYMOUSLY_OWNED = "writer anonymously owned"

class RwsemState(collections.namedtuple):
    state: RwsemStateCode
    owner: drgn.Object
    """A ``struct rwsem_waiter *``, non-NULL for WRITER_OWNED"""
```

Then we could have a single helper that can return all the necessary details. With that, you would be able to see on the command line that the rwsem was anonymously owned, but it wouldn't interfere with other helpers using that helper. In fact, it would make that info available to those helpers as well!

This is just an example, but I think something inspired by this might be good. I'm not opposed to having the print statements in drgn-tools, but I'm not confident they'll be acceptable in drgn upstream (I'm starting to look at that PR as well).

Reply:
I agree @brenns10 and have modified get_rwsem_owner now. Interactive users should use get_rwsem_info but if they are just interested in owner info they can still use get_rwsem_owner as per its description. Let me know how it looks now. And thanks for looking into the upstream PR as well :)

Reply:
I checked https://github.com/oracle-samples/drgn-tools/pull/69/commits/cd9c0e2dfc1f7d6b6b80e832dc06528e3cd4092e and that's great! Just what I was thinking and I think it makes the code a bit clearer.

Just double checking, is this ready for review and merge? I will take a deep look if so.

Reply:
@brenns10 yes its ready for review and merge now